### PR TITLE
refactoring and minor fixes related to logging

### PR DESCRIPTION
**Pull request by ashitsalesforce**
- Fixed a bug introduced in install.bat due to an incorrect use of doublequotes in string substitution.

- Refactoring to use the same log config, config properties, and sample files and directories for supported platforms.

- Fixed a bug in logging during testing so that the integration testing uses test/resources/testfiles/conf/log-conf.xml

- Allow batch mode from DataloaderRunner's main method.

- Fix log4j2 related minor bugs in log-conf.xml

- Migrate test log-conf.xml to log4j2 syntax and make sure that testing code uses test log-conf.xml.

- set the config directory location through java system property if provided as a command line argument to dataloader.command or dataloader.bat to make sure that log-conf.xml gets used when logger is set as a static variable. This is the same approach as the one used by process.bat.

- refactor the code to avoid duplication in creating hashmap of argument name-value pairs in Controller.java and ProcessRunner.java.

### CODE REVIEW

File: src/main/java/com/salesforce/dataloader/dyna/BooleanConverter.java [hunk 1]
```diff
@@ -89,16 +89,16 @@ public BooleanConverter(Object defaultValue) {
...
-            return null;
+        if (value == null || value instanceof Boolean) {
+            return value;
         }
-
-        if (value instanceof Boolean) {
-            return (value);
+      
+        String stringValue = value.toString().trim();
+        if (stringValue.length()==0) {
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/dyna/BooleanConverter.java:97]: you can just do `value.toString().isEmpty()`

File: src/main/java/com/salesforce/dataloader/process/DataLoaderRunner.java [hunk 2]
```diff
@@ -34,26 +34,37 @@
...
 import com.salesforce.dataloader.ui.UIUtils;
 
+import java.util.Map;
+
 import org.eclipse.swt.widgets.Display;
 import org.eclipse.swt.widgets.Shell;
 
 public class DataLoaderRunner {
 
     public static final String UI = "ui";
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/process/DataLoaderRunner.java:44]: Are these variables used outside of this class? Can we narrow the scope if we can?

File: pom.xml [hunk 3]
```diff
@@ -29,6 +29,7 @@
     <test.user.restricted>standard@org.com</test.user.restricted>
     <test.password></test.password><!-- use the encrypted password -->
     <test.redirect>false</test.redirect>
+    <testfiles.dir>${project.build.testOutputDirectory}/testfiles</testfiles.dir>
```
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply pom.xml:32]: Given that this variable refers to testfiles sub-directory within test root directory, how about <test.testfiles.directory> to make it both consistent and non-ambiguous?

File: release/configs/log-conf.xml [hunk 4]
```diff
@@ -10,11 +10,11 @@
   </appenders>
   <loggers>
     <logger name="org.apache.log4j.xml" level="warn">
-      <appender-ref ref="fileAppender"/>
+      <AppenderRef ref="fileAppender"/>
```
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply release/configs/log-conf.xml:13]: Right. Log4J2 uses camel-case.

File: release/win/bin/process.bat [hunk 5]
```diff
@@ -44,8 +44,9 @@ IF "%JAVA_HOME%" == "" (
     IF NOT EXIST "%JAVA_HOME%" (
         echo We couldn't find the Java Runtime Environment ^(JRE^) in directory "%JAVA_HOME%". To run process.bat, set the JAVA_HOME environment variable to the directory where the JRE is installed.
     ) ELSE (
-        "%JAVA_HOME%\bin\java" -cp "%EXE_PATH%\..\dataloader-%DATALOADER_VERSION%-uber.jar" -Dsalesforce.config.dir=%1 com.salesforce.dataloader.process.ProcessRunner %PROCESS_OPTION%
+        "%JAVA_HOME%\bin\java" -cp "%EXE_PATH%\..\dataloader-%DATALOADER_VERSION%-uber.jar" com.salesforce.dataloader.process.DataLoaderRunner salesforce.config.dir=%1 run.mode=batch %PROCESS_OPTION%
     )
 )
 
 :end
+exit /b %errorlevel%
```
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply release/win/bin/process.bat:52]: It exits the bat script but does not exit the shell in which it is run per information [here](https://ss64.com/nt/exit.html).

File: release/win/install.bat [hunk 6]
```diff
@@ -31,7 +31,7 @@ echo It uses '%USERPROFILE%\^<relative path^>' as the installation directory if
 echo.
 set /p DIR_NAME=Provide the installation directory [default: dataloader] : || set DIR_NAME=dataloader
 
-if "%DIR_NAME%":~1,1%" == ":" (
```
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply release/win/install.bat:34]: thanks!

File: src/main/java/com/salesforce/dataloader/dyna/BooleanConverter.java (hunk 1 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/dyna/BooleanConverter.java:97]: Good point. However, given that we need to trim the string and use the trimmed substring to convert it to a Boolean later, I will keep it as-is.

File: src/main/java/com/salesforce/dataloader/controller/Controller.java [hunk 7]
```diff
@@ -83,7 +86,7 @@
  */
 public class Controller {
 
-    private static final String LOG_CONF_OVERRIDE = "log-conf.xml";
+    public static final String LOG_CONF_DEFAULT = "log-conf.xml";
```
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/controller/Controller.java:89]: Will move the line to where other log4j attributes are grouped. The constant is also used by TestBase.java to initialize its own logging. Here is the relevant line in TestBase.java:

` String logConfFilePath = Paths.get(getTestConfDir(), Controller.LOG_CONF_DEFAULT).toString();`

File: src/main/java/com/salesforce/dataloader/process/DataLoaderRunner.java (hunk 2 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/process/DataLoaderRunner.java]: good catch. I have made the change and updated the PR.

**Review by jjangsfdc — APPROVED** [review]
LGTM

File: pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashitsalesforce) [reply pom.xml]: done!

**Comment by bgaspic** [comment]
Hi @ashitsalesforce
Please note that the documentation in https://developer.salesforce.com/docs/atlas.en-us.dataLoader.meta/dataLoader/using_the_command_line_interface.htm has never been updated to this commit, although it's been 10 months.
I just upgraded to v53 because of the log4j jndi:ldap security hole. And I have had to go into GIT log in order to figure out that `com.salesforce.dataloader.process.ProcessRunner` needs to be replaced with `com.salesforce.dataloader.process.DataLoaderRunner run.mode=batch`. Many people who are upgrading dataloader in a hurry, will _not_ realize that, unless it is published in the developer docs page.

**Comment by ashitsalesforce** [comment]
Hi @bgaspic , the documentation you are referring to is specifying how to use ProcessRunner class as an attribute for bean's id tag in process-conf.xml. That works as documented. Is it not working for you?

You are correct that the main method in ProcessRunner is removed and the way to start batch mode is the way you state above. Given that process.bat encapsulates batch mode invocation on Windows, this change is not documented. However, it is captured in README.md:

```
Data Loader can be executed in Batch mode on other unsupported platforms using the following command:

java -jar target/dataloader-x.y.z-uber.jar run.mode=batch process.name=<process name> salesforce.config.dir=<path to config dir containing process-conf.xml and config.properties files>

```

Let me know if I missed something.

**Comment by bgaspic** [comment]
Hi @ashitsalesforce
Apologies, you are correct. All the references in the linked documentation refer to the **XML bean**, not to the method of starting the dataloader in batch mode (which is indeed relegated to `process.bat` in that document)
My issue was caused by starting dataloader batch "directly", from a setup in which I already knew the path to the Zulu Java v11+ - and going for `ProcessRunner` directly, but for Dataloader v53 it has to instead be `DataLoaderRunner run.mode=batch`. Since on the documentation page I saw no reference to `DataLoaderRunner`, and saw it was all references to `ProcessRunner`, I thought the docs were not updated. But since those docs refer to the XML config, it is all correct.
(Incidentally, I was sidestepping `process.bat` due to a customer's IT security setup where it was possible to order the addition of Zulu to someone's machine, but the environment variables would still continue to point to Oracle's Java v8, which would also continue to be the first Java in the PATH - and it was a chore to rearrange those environment variables on the users' machines - whereas the default `process.bat` code was then getting stuck on that Java v8, without "reaching" the Zulu folder which did exist... I didn't open any issue around this since it was solvable, but then I missed this transition to `DataLoaderRunner`)

**Comment by ashitsalesforce** [comment]
good point @bgaspic. I have opened issue #372 about this. Please add more details as needed. As a workaround how about a wrapper bat script that you write that sets JAVA_HOME environment variable to JRE 11 or later and invokes process.bat?

**Pull request merged**